subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/TransformerFromCallable.java

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.artifacts.transform;
+
+import org.gradle.api.artifacts.transform.ArtifactTransformDependencies;
+import org.gradle.api.artifacts.transform.PrimaryInput;
+import org.gradle.api.artifacts.transform.Workspace;
+import org.gradle.api.internal.InstantiatorFactory;
+import org.gradle.api.internal.attributes.ImmutableAttributes;
+import org.gradle.internal.UncheckedException;
+import org.gradle.internal.hash.HashCode;
+import org.gradle.internal.isolation.Isolatable;
+import org.gradle.internal.reflect.ClassDetails;
+import org.gradle.internal.reflect.ClassInspector;
+import org.gradle.internal.reflect.PropertyDetails;
+
+import javax.annotation.Nullable;
+import java.io.File;
+import java.lang.annotation.Annotation;
+import java.lang.reflect.Field;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.Callable;
+
+public class TransformerFromCallable extends AbstractTransformer<Callable<List<File>>> {
+
+    private final Method workspaceSetter;
+    private final Method primaryInputSetter;
+
+    public TransformerFromCallable(Class<? extends Callable<List<File>>> implementationClass, Isolatable<Object[]> paramsSnapshot, HashCode secondaryInputsHash, InstantiatorFactory instantiatorFactory, ImmutableAttributes from) {
+        super(implementationClass, paramsSnapshot, secondaryInputsHash, instantiatorFactory, from);
+        // TODO: make this more general by hooking into `DefaultPropertyMetadataStore` or something similar
+        this.workspaceSetter = findSetterAnnotatedWith(implementationClass, Workspace.class);
+        this.primaryInputSetter = findSetterAnnotatedWith(implementationClass, PrimaryInput.class);
```

Comment:
Two things here:

1) I think we should support configuration injection for these parameters on both kinds of transformer.
2) I think this should work in a way that the user only declares a `Provider<Directory>` or `Provider<RegularFile>` — we shouldn't be using `File` and setters.